workflows/pipe-common/pipeline/hpc/gridengine.py

```diff
@@ -484,3 +546,226 @@ def validate(self, jobs):
+
+    _NODE_BAD_STATES = ["DOWN", "DRAINING", "DRAIN", "DRAINED", "FAIL", "FAILING",  "INVAL"]
+
+    _SCONTROL_DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S'
+    _GET_JOBS = "scontrol -o show job"
+
+    def __init__(self, cmd_executor):
+        self.cmd_executor = cmd_executor
+
+    def get_jobs(self):
+        try:
+            output = self.cmd_executor.execute(SlurmGridEngine._GET_JOBS)
+        except ExecutionError:
+            Logger.warn('Slurm jobs listing has failed.')
+            return []
+        return self._parse_jobs(output)
+
+    def disable_host(self, host):
+        self.cmd_executor.execute(SlurmGridEngine._SCONTROL_UPDATE_NODE_STATE % ("DRAIN", host))
+
+    def enable_host(self, host):
+        host_state = self._get_host_state(host)
+        if "DRAIN" in host_state:
+            self.cmd_executor.execute(SlurmGridEngine._SCONTROL_UPDATE_NODE_STATE % ("UNDRAIN", host))
+        else:
+            # NO-OP for all node states except DRAIN
+            pass
+
+    def get_pe_allocation_rule(self, pe):
+        raise AllocationRuleParsingError("Slurm doesn't have PE preference.")
+
+    def delete_host(self, host, skip_on_failure=False):
+        _perform_command(
+            action=lambda: self.disable_host(host),
+            msg='Disabling GE host execution daemon.',
+            error_msg='Disabling GE host execution daemon has failed.',
+            skip_on_failure=skip_on_failure
+        )
+        _perform_command(
+            action=lambda: self.cmd_executor.execute(SlurmGridEngine._SCONTROL_DELETE_NODE % host),
+            msg='Disabling GE host execution daemon.',
+            error_msg='Disabling GE host execution daemon has failed.',
+            skip_on_failure=skip_on_failure
+        )
+
+    def get_host_supplies(self):
+        for line in self.cmd_executor.execute_to_lines(SlurmGridEngine._SHOW_EXECUTION_HOST % ''):
+            if "NodeName" in line:
+                node_desc = self._parse_dict(line)
+                yield ResourceSupply(cpu=int(node_desc.get("CPUTot", "0"))) \
+                    - ResourceSupply(cpu=int(node_desc.get("CPUAlloc", "0")))
+
+    def get_host_supply(self, host):
+        for line in self.cmd_executor.execute_to_lines(SlurmGridEngine._SHOW_EXECUTION_HOST % host):
+            if "NodeName" in line:
+                node_desc = self._parse_dict(line)
+                return ResourceSupply(cpu=int(node_desc.get("CPUTot", "0"))) \
+                    - ResourceSupply(cpu=int(node_desc.get("CPUAlloc", "0")))
+        return ResourceSupply()
+
+    def get_engine_type(self):
+        return GridEngineType.SLURM
+
+    def is_valid(self, host):
+        node_state = self._get_host_state(host)
+        for bad_state in SlurmGridEngine._NODE_BAD_STATES:
+            if bad_state in node_state:
+                return False
+        return True
+
+    def kill_jobs(self, jobs, force=False):
+        job_ids = [str(job.root_id) for job in jobs]
```

Comment:
Need only unique root ids here.
```suggestion
        job_ids = set(str(job.root_id) for job in jobs)
```